pkgs/native_assets_builder/lib/src/build_runner/build_runner.dart

```diff
@@ -27,15 +27,13 @@ class NativeAssetsBuildRunner {
     required this.dartExecutable,
   });
 
-  final _metadata = <Target, DependencyMetadata>{};
-
   /// [workingDirectory] is expected to contain `.dart_tool`.
   ///
   /// This method is invoked by launchers such as dartdev (for `dart run`) and
   /// flutter_tools (for `flutter run` and `flutter build`).
   ///
   /// Completes the future with an error if the build fails.
-  Future<List<Asset>> build({
+  Future<BuildResult> build({
```

Comment:
I think the `BuildResult` should be per-package. Our flutter or dart tooling should provide nice error messages:

```
[...] Building native packages
[...] Successfully built native libraries a.so, b.so, c.so for package X
[...] Building native package Y had errors:
[...]    Error A
[...]    Error B
[...]    To reproduce, run: ....
```

One could even think about making this API in a streaming fashion, so it can print such things while it's doing things (with the right timestamps) instead of possibly taking minutes and then emitting all this information

Reply:
The sub-tools in dartdev typically do their own error handling and logging, only returning an exit code from the command invocation.

+1 for trying to keep the experience across Dart / Flutter consistent whenever possible (see https://github.com/dart-lang/sdk/issues/52747 which is tracking existing inconsistencies between the tools).

Reply:
+1 to what ben said

Reply:
Okay, so the `BuildResult` should just have a `success` bool field instead of a list of errors. And all the errors should be streamed to the logger.

Thanks for your input @bkonyi and @christopherfujino!

Reply:
> That would duplicate aggregation logic in both flutter_tools and dartdev.  
> ...
> If anything it would be nice if we can print the same messages in both dartdev and flutter_tools. 

Nothing has to be duplicated, shared logic can live in the shared package. This is an API design question - whether logic that does N package builds should return N package build results or not.

Given this isn't a user facing API but only for our own tools and we can change it in the future - it doesn't matter much, so lgtm